### PR TITLE
client/fingerprint: detect unloaded dynamic bridge kernel module

### LINKED ISSUE #8423: Connect jobs failing in 12.0
When I upgraded my cluster to 12.0, all of my Connect jobs began to fail.  I started messing around with syntax of the job definition, but couldn't get it to work.

####  Nomad Version

Nomad v0.12.0 (8f7fbc8e7b5a4ed0d0209968faf41b238e6d5817)

#### OS
Centos 7 and Ubunto 20.04

### Issue
After upgrade to nomad 12.0, allocations fail because of a "missing network" constraint.

### Reproduction Steps

To take any environmental variables out of the mix, I followed the steps outlined on (https://www.nomadproject.io/docs/integrations/consul-connect) exactly.  The only difference between the two runs was the version of the nomad binary running in -dev-connect

#### "nomad job plan connect.nomad"   -   Nomad v0.11.3 (8918fc804a0c6758b6e3e9960e4eb2e605e38552)

+ Job: "countdash"
+ Task Group: "api" (1 create)
  + Task: "connect-proxy-count-api" (forces create)
  + Task: "web" (forces create)

+ Task Group: "dashboard" (1 create)
  + Task: "connect-proxy-count-dashboard" (forces create)
  + Task: "dashboard" (forces create)

Scheduler dry-run:
- All tasks successfully allocated.

Job Modify Index: 0
To submit the job with version verification run:

nomad job run -check-index 0 connect.nomad  -  

When running the job with the check-index flag, the job will only be run if the
job modify index given matches the server-side version. If the index has
changed, another user has modified the job and the plan's results are
potentially invalid.
For reporting security vulnerabilities [please refer to the website.](https://www.nomadproject.io/security.html)

#### "nomad job plan connect.nomad" - Nomad v0.12.0 (8f7fbc8e7b5a4ed0d0209968faf41b238e6d5817)

+ Job: "countdash"
+ Task Group: "api" (1 create)
  + Task: "connect-proxy-count-api" (forces create)
  + Task: "web" (forces create)

+ Task Group: "dashboard" (1 create)
  + Task: "connect-proxy-count-dashboard" (forces create)
  + Task: "dashboard" (forces create)

Scheduler dry-run:
- WARNING: Failed to place all allocations.
  Task Group "api" (failed to place 1 allocation):
    * Constraint "missing network": 1 nodes excluded by filter

  Task Group "dashboard" (failed to place 1 allocation):
    * Constraint "missing network": 1 nodes excluded by filter

Job Modify Index: 0
To submit the job with version verification run:

nomad job run -check-index 0 connect.nomad

When running the job with the check-index flag, the job will only be run if the
job modify index given matches the server-side version. If the index has
changed, another user has modified the job and the plan's results are
potentially invalid.

### PR DESCRIPTION
**Pull request by shoenig**
In Nomad v0.12.0, the client added additional fingerprinting around the
presense of the bridge kernel module. The fingerprinter only checked in
`/proc/modules` which is a list of loaded modules. In some cases, the
bridge kernel module is builtin rather than dynamically loaded. The fix
for that case is in #8721. However we were still missing the case where
the bridge module is dynamically loaded, but not yet loaded during the
startup of the Nomad agent. In this case the fingerprinter would believe
the bridge module was unavailable when really it gets loaded on demand.

This PR now has the fingerprinter scan the kernel module dependency file,
which will contain an entry for the bridge module even if it is not yet
loaded.

In summary, the client now looks for the bridge kernel module in
 - `/proc/modules`
 - `/lib/modules/<kernel>/modules.builtin`
 - `/lib/modules/<kernel>/modules.dep`

Closes #8423

### CODE REVIEW

**Comment by shoenig** [comment]
This is trying hard to avoid an `exec modinfo`; open to weighing the pros and cons.

**Review by tgross — APPROVED** [review]
LGTM 👍 

I took a look at the source of [`modprobe`](https://git.kernel.org/pub/scm/utils/kernel/module-init-tools/module-init-tools.git/tree/modprobe.c) and the output of strace while searching for a bogus module and it looks like this should work.

One small caveat is that modprobe has configuration values that can tell it to look elsewhere; in practice I suspect this isn't something that's going to impact `bridge` but we should at least have that in the back of our mind if we get a report about someone's exotic setup later down the road.

**Comment by apollo13** [comment]
Hi @shoenig, on which systems did you check this PR? @jeroentbt ran over this in Gitter and told me that he uses CentOS. When I checked against my CentOS 7 boxes they do have the following in `modules.dep`:
```
[root@mon01 ~]# cat /lib/modules/3.10.0-1160.11.1.el7.x86_64/modules.dep|grep bridge.ko.xz
kernel/drivers/net/ethernet/mellanox/mlxsw/mlxsw_spectrum.ko.xz: kernel/drivers/net/ethernet/mellanox/mlxfw/mlxfw.ko.xz kernel/net/ipv6/ip6_tunnel.ko.xz kernel/net/ipv6/tunnel6.ko.xz kernel/net/psample/psample.ko.xz kernel/lib/parman.ko.xz kernel/net/bridge/bridge.ko.xz kernel/net/802/stp.ko.xz kernel/net/llc/llc.ko.xz kernel/drivers/net/ethernet/mellanox/mlxsw/mlxsw_pci.ko.xz kernel/drivers/net/ethernet/mellanox/mlxsw/mlxsw_core.ko.xz kernel/net/core/devlink.ko.xz
kernel/drivers/media/pci/ddbridge/ddbridge.ko.xz: kernel/drivers/media/dvb-core/dvb-core.ko.xz
kernel/net/netfilter/xt_physdev.ko.xz: kernel/net/bridge/br_netfilter.ko.xz kernel/net/bridge/bridge.ko.xz kernel/net/802/stp.ko.xz kernel/net/llc/llc.ko.xz
kernel/net/bridge/netfilter/nf_tables_bridge.ko.xz: kernel/net/netfilter/nf_tables.ko.xz kernel/net/netfilter/nfnetlink.ko.xz
kernel/net/bridge/netfilter/nft_meta_bridge.ko.xz: kernel/net/netfilter/nft_meta.ko.xz kernel/net/netfilter/nf_tables.ko.xz kernel/net/netfilter/nfnetlink.ko.xz
kernel/net/bridge/netfilter/nft_reject_bridge.ko.xz: kernel/net/bridge/netfilter/nf_tables_bridge.ko.xz kernel/net/ipv4/netfilter/nf_reject_ipv4.ko.xz kernel/net/ipv6/netfilter/nf_reject_ipv6.ko.xz kernel/net/netfilter/nft_reject.ko.xz kernel/net/netfilter/nf_tables.ko.xz kernel/net/netfilter/nfnetlink.ko.xz kernel/net/bridge/bridge.ko.xz kernel/net/802/stp.ko.xz kernel/net/llc/llc.ko.xz
kernel/net/bridge/netfilter/nf_log_bridge.ko.xz:
kernel/net/bridge/netfilter/ebtable_broute.ko.xz: kernel/net/bridge/netfilter/ebtables.ko.xz kernel/net/bridge/bridge.ko.xz kernel/net/802/stp.ko.xz kernel/net/llc/llc.ko.xz
kernel/net/bridge/bridge.ko.xz: kernel/net/802/stp.ko.xz kernel/net/llc/llc.ko.xz
kernel/net/bridge/br_netfilter.ko.xz: kernel/net/bridge/bridge.ko.xz kernel/net/802/stp.ko.xz kernel/net/llc/llc.ko.xz
```

would it be possible to include `(\.xz)?` in the regex in https://github.com/hashicorp/nomad/pull/9299/files#diff-25979a0adb76fadc59ae57997b3d3e70ac415168d760bd3e6d95b8db1cca7cc6R19 ?

**Comment by shoenig** [comment]
Thanks for pointing that out @apollo13, I think fixing the regexp makes sense. I created https://github.com/hashicorp/nomad/issues/9776 to track that. Would you like to open a PR?
